design-proposals/002-filter-data.md

```diff
@@ -0,0 +1,142 @@
+* Affected Repos:
+  - syndesis-rest
+  - syndesis-ui
+
+### User Story
+
+As a citizen user, I would like to restrict the data that flows through an integration to match a specific set of criteria.  For example, I may have an integration that is triggered every time a tweet mentions my account, but I only want to process tweets that contain the text "Syndesis".
+
+I have two initial requirements for filtering data in an integration:
+* Based on content in the body.
+* Based on content in a message header.
+
+In both cases, I expect to be able to apply basic string operations to the data to form a filter expression (contains, does not contain, starts with).  For more details on options, see the Operator Support section here:
+http://camel.apache.org/simple.html
+
+Similar to logging, the UI should help in creating filter expressions for new/inexperienced users (like through clickable examples or typed form inputs).  The interface should also allow for advanced users to enter free form expressions.
+
+### Domain
+
+#### UI defined filter (typed)
+
+For the typed definition of a filter, a **filter step** contains one or more **filter rules** which are combined with _and_ or _or_
+
+A _filter step_ is a regular `Step` with a type "filter". 
+_Filter rules_ are part of the configuration of this step and stored in the properties of a step.
+The lowest building block is a a `FilterRule`. It has the following properties:
+
+* **type** : On which part of the message to apply the filter. Can be either "header", "body" or "properties".
+* **key** : The key to use to extract a message value (e.g. header or property name). The key is empty when type is equals to body.
+* **op** : The operator to use for the filter
+* **value** : The value used by the operator to determine whether a filter applies.
+* **combinator** : How this rule is combined with the previous rule. The value can be either `&&` or `||`. The first filter rule's combinator is ignored when building up the expression.
+ 
+"type" and "key" could also be combined to a selector (which e.g. then would have the format `type.key`)
+
+The configuration of a `FilterStep` is then a list of `FilterRules`.
+
+For example, the following configuration
+
+```yaml
+- type: "body"
+  op: "contains"
+  value: "antman"
+- type: "header"    
+  key: "region"
+  op: "=~"
+  value: "asia"
+  combinator: "&&"
+- type: "body"
+  op: "regex"
+  value: "bat(wo)?man"
+  combinator: "||"
+- type: "header"
+  key: "publisher"
+  op: "=~"
+  value: "DC Comics"
+  combinator: "&&"
+```
+
+would be stored as a property "filter" on the integration step. It translates later to simple lang expression
+
+```
+${body} contains "antman" && ${in.header.region} =~ "asia" || ${body} regex "bat(wo)?man" && ${in.header.publisher} =~ "DC Comics"
+```
+
+The simple expression language [does not support parentheses](http://camel.apache.org/simple.html) nor precedence of operators so the expression is always evaluated from left to right.
```

Comment:
Good question. From an UI POV it would be simpler to have only a single expression. For an expanded use case one could then go to the text field. I wouldn't mind to omit 'and' and 'or' for now.